### ISSUE TITLE
Add instructions for suspend/hibernate with an encrypted swap partition and/or file

### ISSUE THREAD

**Issue by seanthegeek** [wmutschl/mutschler.eu#13]
The Ubuntu 20.04 btrfs-luks guide [states](https://mutschler.eu/linux/install-guides/ubuntu-btrfs/#encrypted-swap):

> As I have no use for hibernation or suspend-to-disk, I will simply use a random password to decrypt the swap partition using the crypttab

For my use case, I need to have a fully encrypted laptop with suspend-to-disk/hibernate support. How can I do this? The [section about it](https://wiki.archlinux.org/index.php/Dm-crypt/Swap_encryption#With_suspend-to-disk_support) in dm-crypt/Swap encryption is not clear to me.

**Comment by noyannus** [comment]
[That](https://forum.manjaro.org/t/hibernation-not-working-need-help/66539) might have some useful info.

**Comment by wmutschl** [comment]
Once I find some time, I will update the guides for 21.04 and will show how to enable hibernation. In Pop!_OS it is very easy by simply encrypting the swap partition in your crypttab. I am not sure whether this will also work in Ubuntu and what changes are required in Grub. I will leave this issue open until I update the guides.

**Comment by DeadSeeds** [comment]
Same problem here: I can't get hibernation to work properly.  I've done a ton of research, but can't seem to get the kernelstub or crypttab configuration correct.  When I try to hibernate, journalctl reports PM: Image not found (code -22).  My encrypted swap is working correctly (I've tested it) for everything but hibernation.  I guess the hibernation file is never being written?  Not sure.  Either way, your help getting this working would be very much appreciated.  And thank you for the great BTRFS + LUKS + PopOS tutorial!  I learned a lot.

**Comment by hluengas** [comment]
+1 on this, 21.04 guid for luks+btrfs worked great but I can't get the hibernation to work as guides are all over the place online

**Comment by wmutschl** [comment]
Fyi there is also a lengthy discussion on the discourse of Ubuntu [Re-visiting hibernate on ubuntu](https://discourse.ubuntu.com/t/re-visiting-hibernate-on-ubuntu/15953).

**Comment by RYcWT46Lsb5awE** [comment]
I spent a lot of time trying to find a solution to this. Hopefully one day I'll get around to making polished instructions in a PR, but here's an info dump for now:

**TL;DR swap partition encrypted with a /dev/urandom key will not support hibernation.** In theory you could make a keyscript to make it work by saving the random key on disk somewhere, but there are literally zero benefits to this approach. Just use your data partition password (or a separate password) to unlock swap directly.

_For any auto-login users, consider using `keyscript=decrypt_keyctl` as in method 2. Then you can unlock your keyring with the encryption password._
[Source](https://www.reddit.com/r/debian/comments/kw2cl2/comment/gj6mmyj)

## Method 1: Separate Password
_Note that you will have to enter two passwords on every startup (from shutdown or hibernate)._

1. Create a LUKS partition for swap with a password of choice:
```
cryptsetup luksFormat /dev/nvme0n1p3
```
2. Change the line for the swap partition in `/etc/crypttab` to look like this:
```
cryptswap UUID=bla-bla-bla none luks,discard,swap
```
3. Add `"resume=/dev/mapper/cryptswap"` as a kernel flag:
```
sudo kernelstub -a "resume=/dev/mapper/cryptswap"
```
(or manually add it to the "user" section of `/etc/kernelstub/configuration`)

4. Add line `RESUME=/dev/mapper/cryptswap` to `/etc/initramfs-tools/conf.d/resume`, or replace `RESUME` line if it exists.

5. Update initramfs:
```
update-initramfs -c -k all
```
[Source 1](https://help.ubuntu.com/community/EnableHibernateWithEncryptedSwap), [Source 2](https://abskmj.github.io/notes/posts/pop-os/enable-hibernate/), [Source 3](https://gist.github.com/xfreshmanx/74af7d98cc1fc918eb5bbb7f91907099)

## Method 2 (Recommended): Kernel Keyring
Feel free to read the top of `/usr/lib/cryptsetup/scripts/decrypt_keyctl` or the linked source for an explanation. Basically, your LUKS password(s) can be cached in the kernel keyring and used to unlock other partitions or [even unlock the keyring](https://www.reddit.com/r/debian/comments/kw2cl2/comment/gj6mmyj).
1. Create a LUKS partition for swap **with the same password as cryptdata**:
```
cryptsetup luksFormat /dev/nvme0n1p3
```
2. Change `/etc/crypttab` to look like this:
```
cryptdata UUID=something-something none luks,discard,keyscript=decrypt_keyctl
cryptswap UUID=bla-bla-bla none luks,discard,swap,tries=1,keyscript=decrypt_keyctl
```
3. Add `"resume=/dev/mapper/cryptswap"` as a kernel flag:
```
sudo kernelstub -a "resume=/dev/mapper/cryptswap"
```
(or manually add it to the "user" section of `/etc/kernelstub/configuration`)

4. Add line `RESUME=/dev/mapper/cryptswap` to `/etc/initramfs-tools/conf.d/resume`, or replace `RESUME` line if it exists.

5. Update initramfs:
```
update-initramfs -c -k all
```
[Source](https://github.com/gebi/keyctl_keyscript) plus sources from method 1

## Method 3: Swap as LVM Volume
I have not tried this at all, but it should be possible to create a swap partition inside cryptdata using `lvcreate -C`. That way only one cryptdevice needs to be unlocked to decrypt both data and swap. Of course you will need to specify UUIDs rather than `/dev/mapper/cryptswap` in `/etc/fstab`, the kernelstub configuration, and `/etc/initramfs-tools/conf.d/resume`.
[Source](https://unix.stackexchange.com/a/144597)

---

As an aside, my primary motivation for this was to ensure nothing is left decrypted after a period of extended inactivity. That is, I want the system to always hibernate after sitting in suspend for some time. Creating a symlink `/etc/systemd/system/systemd-suspend.service` that points to `/usr/lib/systemd/system/systemd-suspend-then-hibernate.service` overrides the normal suspend mode with suspend-then-hibernate, no matter how the suspend is triggered (power menu, inactivity, command line, etc.). `HibernateDelaySec` in `/etc/systemd/sleep.conf` controls how long the system will stay in suspend before hibernating (you can set it in minutes like `HibernateDelaySec=30min`). Of course this behavior is not necessary for working hibernation, but it will ensure the system can't stay decrypted (and running) indefinitely.
[Source 1](https://askubuntu.com/a/1316753), [Source 2](https://medium.com/@gayanper/make-ubuntu-sleep-like-windows-4761a91f62c2)

**Comment by wmutschl** [comment]
Thank you for this; this should provide ample information for those who want to try out the different methods.
